src/common/types.cpp

```diff
@@ -1180,6 +1211,41 @@ const LogicalType &MapType::ValueType(const LogicalType &type) {
 	return ListType::GetChildType(StructType::GetChildTypes(type)[1].second);
 }
 
+//===--------------------------------------------------------------------===//
+// Union Type
+//===--------------------------------------------------------------------===//
+
+LogicalType LogicalType::UNION(child_list_t<LogicalType> members) {
+	// union types always have a hidden "tag" field in front
+	members.insert(members.begin(), {"", LogicalType::TINYINT});
+	auto info = make_shared<StructTypeInfo>(move(members));
+	return LogicalType(LogicalTypeId::UNION, move(info));
+}
+
+const LogicalType &UnionType::GetMemberType(const LogicalType &type, idx_t index) {
+	auto &child_types = StructType::GetChildTypes(type);
+	D_ASSERT(index < child_types.size());
+	// skip the "tag" field
+	return child_types[index + 1].second;
+}
+
+const string &UnionType::GetMemberName(const LogicalType &type, idx_t index) {
+	auto &child_types = StructType::GetChildTypes(type);
+	D_ASSERT(index < child_types.size());
+	// skip the "tag" field
+	return child_types[index + 1].first;
+}
+
+idx_t UnionType::GetMemberCount(const LogicalType &type) {
+	// dont count the "tag" field
+	return StructType::GetChildTypes(type).size() - 1;
+}
+const child_list_t<LogicalType> UnionType::GetMemberTypes(const LogicalType &type) {
```

Comment:
Maybe check if this can be fully removed? If not that is also fine

Reply:
I've changed almost every use case where it makes sense to use the index accessors instead. There are still some places where we actually want to copy the vector where it remains, but I've renamed the function to `UnionType::CopyMemberTypes()` to make it more explicit